src/lightning_lite/strategies/strategy.py

```diff
@@ -314,3 +304,18 @@ def no_backward_sync(self, module: Module) -> Generator:
 
         This is a context manager. It is only effective if it wraps a call to `.backward()`.
         """
+
+
+class _Sharded(ABC):
```

Comment:
can we make this a protocol instead? Then strategy authors don't need to inherit from a private baseclass but instancechecks would still work within our codebase.

Reply:
My intention here was to make it very clear what the different interfaces are that a strategy exposes. Today, it is just "a bunch of methods that you can call". With the protocol, it is not as clear, because the interface wouldn't be listed as parent in the class definition. 

Also, this is just protected here for now. We are the only strategy authors at the moment. If I had made it public, someone would have requested to make it not public.

Reply:
To me it's less about public or private.

To me this sounds like you want to check if a strategy implements a certain thing and this is what Protocols are for in python. 

With abstract classes, you force strategies to implement this as soon as they inherit from it (which also is a valid approach).

So I guess it comes down to the following statement:
Personally, I don't like multiple baseclasses as long as there is an alternative (which here we have, as  it is just about API definition/check and not about actually providing functionality).

I am fine either way (which is why I also approved), that's just a preference. So feel free to resolve it if that's the only blocker left :)

Reply:
I agree that multiple inheritance is not the right choice in many cases, and can often be avoided for a better design. In this case, I believe it does have a valid and strong use case. The difference to "just adding a base class" is that here it is specifically purely an interface (all methods are abstract). No implementation is provided. I would definitely not do this if I had to provide an implementation to the `_Sharded` method(s).

> With abstract classes, you force strategies to implement this as soon as they inherit from it (which also is a valid approach).

This is 100% what I was going for. My requirements that guided the choice here were:
- We want to go towards flattening the strategy inheritance hierarchy
- When flattening, avoid adding methods to the base strategy class while they only apply to certain strategy types
- Provide better documentation which methods are required to be implemented and which ones are optional.

I believe only the current approach can satisfy all these requirements.

Does @carmocca have any strong opinions about this? Otherwise I would keep it.

Reply:
I mostly agree with Adrian. It's true we could go the protocol way here and things would work, but why would we when we control all these implementations. So far we've only used protocols to fill the gaps in the designs of PyTorch, which we don't control